README.md
[![Stories in Ready](https://badge.waffle.io/typo3-coreapi/ext-coreapi.png?label=ready&title=Ready)](https://waffle.io/typo3-coreapi/ext-coreapi) [![Build Status](https://travis-ci.org/TYPO3-coreapi/ext-coreapi.svg?branch=feature%2FMakeExtensionApiCompatibleTo62)](https://travis-ci.org/TYPO3-coreapi/ext-coreapi) [![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/TYPO3-coreapi/ext-coreapi/badges/quality-score.png?b=feature%2FMakeExtensionApiCompatibleTo62)](https://scrutinizer-ci.com/g/TYPO3-coreapi/ext-coreapi/?branch=feature%2FMakeExtensionApiCompatibleTo62) [![Code Coverage](https://scrutinizer-ci.com/g/TYPO3-coreapi/ext-coreapi/badges/coverage.png?b=feature%2FMakeExtensionApiCompatibleTo62)](https://scrutinizer-ci.com/g/TYPO3-coreapi/ext-coreapi/?branch=feature%2FMakeExtensionApiCompatibleTo62)

## TYPO3 Extension 'coreapi'

The EXT:coreapi should provide a simple to use API for common core features. Goal is to be able to do the most common tasks by CLI instead of doing it in the backend/browser.

Beside of CLI commands, EXT:coreapi provides service classes, which can be used in your own implementation/extension.

Checkout the project website at forge.typo3.org:
	http://forge.typo3.org/projects/show/extension-coreapi

### Tasks
* DatabaseApi
	* databaseCompare
* CacheApi
	* clearAllCaches
	* clearPageCache
	* clearConfigurationCache
* ExtensionApi
	* info
	* listInstalled
	* updateList from TER
	* fetch an extension from TER
	* import an extension
	* install / uninstall extension
	* ~~create upload folders~~
	* configure extension
* SiteApi
	* info
	* createSysNews

#### planned/comming soon

* Backend
	* manage users (list, create, update, delete)
	* lock/unlock the TYPO3 backend
* PageTree
	* print/get
* DataApi
 	* generic list/create/update/delete records (and not doing the plain SQL, but using the DataHandler (aka tcemain)!)
	* getRecordsByPid
	* create a database dump (exclude "temporary" tables like caches, sys_log, ...)
* ReportsApi
	* run/check the reports from the reports module
* ConfigurationApi
	* list, get and set TYPO3 configurations


### CLI call: ###

Make sure you have a backend user called `_cli_lowlevel`

If you want to use the cache clearing commands, you need to add the following snippet to the TSconfig field of this backend user:

	options.clearCache.all=1
	options.clearCache.pages=1

#### TYPO3 6.2 and below ####
Support for TYPO3 CMS below 6.2 was removed with version 0.2.0 of this extension. In case you need to use ext:coreapi in combination with lower version of TYPO3 CMS use version [0.1.0-beta](https://github.com/TYPO3-coreapi/ext-coreapi/releases/tag/0.1.0-beta "0.1.0-beta").

#### TYPO3 6.2+ ####
If you are using TYPO3 6.2+, you can use the awesome CommandController of Extbase.

This will show you all available calls
	./typo3/cli_dispatch.phpsh extbase help

### Usage in Composer ###

    {
        "name": "typo3cms/test-website",
        "description": "TYPO3 CMS: test.com",
        "keywords": ["typo3", "cms"],
        "require": {
            "php": ">=5.3.3",
            "typo3core/cms": "*",
            "etobi/coreapi": "dev-master",
        },
        "extra": {
            "installer-paths": {
                "typo3conf/ext/{$name}": [
                    "type:typo3-cms-extension"
                ]
            }
        },
        "minimum-stability": "dev",
        "require-dev": {},
        "scripts": {}
    }

### Running the unit tests

The Unit Tests rely on [vfsStream](https://github.com/mikey179/vfsStream "vfsStream"). For some reasons ext:coreapi don't add this dependencies by itself but uses the one which is alread defined for Core Unit Tests.
To install vfsStream copy the composer.json from the TYPO3 CMS package into you webroot folder and execute the command `composer install`. This will install all dependencies into Packages/Libraries/.

Then run the Unit Tests.

    cp typo3_src/composer.json .
    composer install
    ./bin/phpunit --colors -c typo3/sysext/core/Build/UnitTests.xml typo3conf/ext/coreapi/Tests/Unit/


CHANGELOG.md
# Change Log

## [1.2.1](https://github.com/TYPO3-coreapi/ext-coreapi/tree/1.2.1) (2015-11-15)

[Full Changelog](https://github.com/TYPO3-coreapi/ext-coreapi/compare/1.2.0...1.2.1)

**Closed issues:**

- extensionapi:info throws error [\#108](https://github.com/TYPO3-coreapi/ext-coreapi/issues/108)

- Add classloading information to composer.json [\#105](https://github.com/TYPO3-coreapi/ext-coreapi/issues/105)

**Merged pull requests:**

- Add autoloading [\#111](https://github.com/TYPO3-coreapi/ext-coreapi/pull/111) ([DanielSiepmann](https://github.com/DanielSiepmann))

- \[BUGFIX\] Changes self::MAXIMUM\_LINE\_LENGTH to -\>output-\>getMaximumLineLength\(\) [\#110](https://github.com/TYPO3-coreapi/ext-coreapi/pull/110) ([jmverges](https://github.com/jmverges))

- Fixed bug  [\#109](https://github.com/TYPO3-coreapi/ext-coreapi/pull/109) ([thomashohn](https://github.com/thomashohn))

- \[TASK\] Add autoload information to composer.json [\#107](https://github.com/TYPO3-coreapi/ext-coreapi/pull/107) ([smichaelsen](https://github.com/smichaelsen))

- \[BUGFIX\] missing constants MAXIMUM\_LINE\_LENGTH added [\#106](https://github.com/TYPO3-coreapi/ext-coreapi/pull/106) ([ddiebel](https://github.com/ddiebel))

## [1.2.0](https://github.com/TYPO3-coreapi/ext-coreapi/tree/1.2.0) (2015-05-05)

[Full Changelog](https://github.com/TYPO3-coreapi/ext-coreapi/compare/1.1.1...1.2.0)

**Implemented enhancements:**

- Add command which deletes really all caches [\#96](https://github.com/TYPO3-coreapi/ext-coreapi/issues/96)

**Closed issues:**

- Adding a Command to clear the cache like it is possible to do in the install tool [\#90](https://github.com/TYPO3-coreapi/ext-coreapi/issues/90)

- Add HHVM config [\#114](https://github.com/TYPO3-coreapi/ext-coreapi/issues/114)

- Add HHVM as build environment on Travis-CI and allow it to fail [\#113](https://github.com/TYPO3-coreapi/ext-coreapi/issues/113)

- Migrate to container-based infrastructure on Travis-CI [\#112](https://github.com/TYPO3-coreapi/ext-coreapi/issues/112)

**Merged pull requests:**

- \[TASK\] Remove leading Slash on class names [\#104](https://github.com/TYPO3-coreapi/ext-coreapi/pull/104) ([7elix](https://github.com/7elix))

- Exception with TYPO3 7 [\#103](https://github.com/TYPO3-coreapi/ext-coreapi/pull/103) ([smichaelsen](https://github.com/smichaelsen))

- \[TASK\] using sprintf for logger message too [\#102](https://github.com/TYPO3-coreapi/ext-coreapi/pull/102) ([greinerd](https://github.com/greinerd))

- \[FEATURE\] Make part of factory default [\#101](https://github.com/TYPO3-coreapi/ext-coreapi/pull/101) ([mbrodala](https://github.com/mbrodala))

- Add requirement for typo3/cms to composer.json [\#98](https://github.com/TYPO3-coreapi/ext-coreapi/pull/98) ([rabe69](https://github.com/rabe69))

- \[FEATURE\] Adds command to create uploads folders of all installed extensions [\#97](https://github.com/TYPO3-coreapi/ext-coreapi/pull/97) ([lars85](https://github.com/lars85))

- \[TASK\] ConfigurationApi [\#89](https://github.com/TYPO3-coreapi/ext-coreapi/pull/89) ([achimfritz](https://github.com/achimfritz))

## [1.1.1](https://github.com/TYPO3-coreapi/ext-coreapi/tree/1.1.1) (2014-10-28)

[Full Changelog](https://github.com/TYPO3-coreapi/ext-coreapi/compare/1.1.0...1.1.1)

## [1.1.0](https://github.com/TYPO3-coreapi/ext-coreapi/tree/1.1.0) (2014-10-28)

[Full Changelog](https://github.com/TYPO3-coreapi/ext-coreapi/compare/1.0.0-beta...1.1.0)

**Implemented enhancements:**

- Refactor calls to t3lib\_div::makeInstance [\#93](https://github.com/TYPO3-coreapi/ext-coreapi/issues/93)

- code cleanup [\#33](https://github.com/TYPO3-coreapi/ext-coreapi/issues/33)

- refactor database compare [\#31](https://github.com/TYPO3-coreapi/ext-coreapi/issues/31)

- Make extensionapi:fetch working with TYPO3 CMS \>= 6.2 [\#67](https://github.com/TYPO3-coreapi/ext-coreapi/issues/67)

**Fixed bugs:**

- databaseapi:databasecompare drops needed 'categories' fields [\#92](https://github.com/TYPO3-coreapi/ext-coreapi/issues/92)

**Closed issues:**

- Release to TER [\#58](https://github.com/TYPO3-coreapi/ext-coreapi/issues/58)

- logging with the new logging api [\#30](https://github.com/TYPO3-coreapi/ext-coreapi/issues/30)

- lock/unlock the TYPO3 backend [\#22](https://github.com/TYPO3-coreapi/ext-coreapi/issues/22)

- return valid exit codes [\#19](https://github.com/TYPO3-coreapi/ext-coreapi/issues/19)

**Merged pull requests:**

- adding option to clear all caches hard like in install tool [\#91](https://github.com/TYPO3-coreapi/ext-coreapi/pull/91) ([greinerd](https://github.com/greinerd))

- Clear all active opcode caches [\#88](https://github.com/TYPO3-coreapi/ext-coreapi/pull/88) ([bgmgmbh](https://github.com/bgmgmbh))

- Clear the new system cache [\#87](https://github.com/TYPO3-coreapi/ext-coreapi/pull/87) ([bgmgmbh](https://github.com/bgmgmbh))

- Clear system cache [\#71](https://github.com/TYPO3-coreapi/ext-coreapi/pull/71) ([bgmgmbh](https://github.com/bgmgmbh))

## [1.0.0-beta](https://github.com/TYPO3-coreapi/ext-coreapi/tree/1.0.0-beta) (2014-07-13)

[Full Changelog](https://github.com/TYPO3-coreapi/ext-coreapi/compare/0.1.0-beta...1.0.0-beta)

**Implemented enhancements:**


Classes/Cli/Dispatcher.php
<?php
/***************************************************************
 *  Copyright notice
 *
 *  (c) 2012 Georg Ringer (georg.ringer@cyberhouse.at)
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

/**
 * Starts all due tasks, used by the command line interface
 * This script must be included by the "CLI module dispatcher"
 *
 * @author Georg Ringer <georg.ringer@cyberhouse.at>
 * @package TYPO3
 * @subpackage tx_coreapi
 */
class Tx_Coreapi_Cli_Dispatcher extends t3lib_cli {

	var $cli_help = array(
		'name' => 'coreapi',
		'synopsis' => './cli_dispatch.phpsh coreapi service:command [options] arguments',
		'description' => '
Coreapi provides a set of services/commands for doing the most common admin task in TYPO3 by CLI instead of doing it in the backend/browser.
Currently the following commands are supported:

###COMMANDS###
',
		'examples' => '
./cli_dispatch.phpsh coreapi site:info
./cli_dispatch.phpsh coreapi cache:clearallcaches
./cli_dispatch.phpsh coreapi extension:info rtehtmlarea
./cli_dispatch.phpsh coreapi site:createsysnews "title" "text"',
		'options' => 'use "./cli_dispatch.phpsh coreapi help service:command" to get help on the arguments and options of a specific command
		',
		'license' => 'GNU GPL - free software!',
		'author' => 'Tobias Liebig',
	);

	const MAXIMUM_LINE_LENGTH = 79;

	var $commandMethodPattern = '/([a-z][a-z0-9]*)([A-Z][a-zA-Z0-9]*)Command/';

	/**
	 * @var string service
	 */
	protected $service = '';

	/**
	 * @var string command
	 */
	protected $command = '';

	/**
	 * Constructor with basic checks
	 */
	public function __construct() {
		parent::__construct();

		if (!isset($this->cli_args['_DEFAULT'][1]) || $this->cli_args['_DEFAULT'][1] === 'help') {
			$this->cli_help();
			die();
		}

		$split = explode(':', $this->cli_args['_DEFAULT'][1]);
		if (count($split) === 1) {
			$this->error('CLI calls need to be like coreapi cache:clearallcaches');
		} elseif (count($split) !== 2) {
			$this->error('Only one : is allowed in first argument');
		}

		$this->service = strtolower($split[0]);
		$this->command = strtolower($split[1]);
	}

	/**
	 * Starts the script
	 * @return void
	 */
	public function start() {
		try {
			$command = $this->service . ucfirst($this->command) . 'Command';
			$this->runCommand($command);

		} catch (Exception $e) {
			$errorMessage = sprintf('ERROR: Error in service "%s" and command "%s"": %s!', $this->service, $this->command, $e->getMessage());
			$this->outputLine($errorMessage);
		}
	}

	/**
	 * @param $command
	 * @throws InvalidArgumentException
	 */
	protected function runCommand($command) {
		if (method_exists($this, $command)) {
			$args = array_slice($this->cli_args['_DEFAULT'], 2);
			$method = new ReflectionMethod(get_class($this), $command);

			//check number of required arguments
			if ($method->getNumberOfRequiredParameters() !== count($args)) {
				throw new InvalidArgumentException('Wrong number of arguments');
			}

			foreach ($method->getParameters() as $param) {
				if ($param->isOptional()) {
					$name = $param->getName();
					if ($this->cli_isArg('--' . $name)) {
						$args[] = $this->cli_argValue('--' . $name);
					} else {
						$args[] = $param->getDefaultValue();
					}
				}
			}
			//invoke command with given args and options
			$method->invokeArgs($this, $args);
		} else {
			throw new InvalidArgumentException('Service does not exist or command not supported');
		}
	}

	/**
	 * Clear all caches
	 *
	 * Clears all TYPO3 caches
	 *
	 * @return void
	 * @example ./cli_dispatch.phpsh coreapi cache:clearallcaches
	 */
	public function cacheClearallcachesCommand() {
		$cacheApiService = $this->getCacheApiService();
		$cacheApiService->clearAllCaches();
		$this->outputLine('All caches cleared');
	}


	/**
	 * Clear configuration cache (temp_CACHED_..)
	 *
	 * Deletes the temp_CACHED_* files in /typo3conf
	 *
	 * @return void
	 * @example ./cli_dispatch.phpsh coreapi cache:clearconfigurationcache
	 */
	public function cacheClearconfigurationcacheCommand() {
		$cacheApiService = $this->getCacheApiService();
		$cacheApiService->clearConfigurationCache();
		$this->outputLine('Configuration cache cleared');
	}

	/**
	 * Clear page cache
	 *
	 * Clears the page cache in TYPO3
	 *
	 * @return void
	 * @example ./cli_dispatch.phpsh coreapi cache:clearpagecache
	 */
	public function cacheClearpagecacheCommand() {
		$cacheApiService = $this->getCacheApiService();
		$cacheApiService->clearPageCache();
		$this->outputLine('Page cache cleared');
	}

	/**
	 * Clear all caches except the page cache.
	 * This is especially useful on big sites when you can't just drop the page cache
	 *
	 * @example ./cli_dispatch.phpsh coreapi cache:clearallexceptpagecache
	 * @return void
	 */
	public function cacheClearallexceptpagecacheCommand() {
		$cacheApiService = $this->getCacheApiService();
		$clearedCaches = $cacheApiService->clearAllExceptPageCache();

		$this->outputLine('Cleared caches: ' . implode(', ', $clearedCaches));
	}

	/**
	 * Database compare
	 *
	 * Leave the argument 'actions' empty or use "help" to see the available ones
	 *
	 * @param string $actions List of actions which will be executed
	 * @return void
	 * @example ./cli_dispatch.phpsh coreapi database:databasecompare 2
	 */
	public function databaseDatabasecompareCommand($actions) {
		$databaseApiService = $this->getDatabaseApiService();
		if ($actions === 'help') {
			$actions = $databaseApiService->databaseCompareAvailableActions();
			$this->outputTable($actions);
		} else {
			$databaseApiService->databaseCompare($actions);
		}
	}

	/**
	 * Information about an extension
	 *
	 * Echo's out a table with information about a specific extension
	 *
	 * @param string $extkey extension key
	 * @return void
	 * @example ./cli_dispatch.phpsh coreapi extension:info rtehtmlarea
	 */
	public function extensionInfoCommand($extkey) {
		$extensionApiService = $this->getExtensionApiService();
		$data = $extensionApiService->getExtensionInformation($extkey);
		$this->outputLine('');
		$this->outputLine('EXTENSION "%s": %s %s', array(strtoupper($extkey), $data['em_conf']['version'], $data['em_conf']['state']));
		$this->outputLine(str_repeat('-', self::MAXIMUM_LINE_LENGTH));

		$outputInformation = array();
		$outputInformation['is installed'] = ($data['is_installed'] ? 'yes' : 'no');
		foreach ($data['em_conf'] as $emConfKey => $emConfValue) {
			// Skip empty properties
			if (empty($emConfValue)) {
				continue;
			}
			// Skip properties which are already handled
			if ($emConfKey === 'title' || $emConfKey === 'version' || $emConfKey === 'state') {
				continue;
			}
			$outputInformation[$emConfKey] = $emConfValue;
		}

		foreach ($outputInformation as $outputKey => $outputValue) {
			$description = '';
			if (is_array($outputValue)) {
				foreach ($outputValue as $additionalKey => $additionalValue) {
					if (is_array($additionalValue)) {

						if (empty($additionalValue)) {
							continue;
						}
						$description .= LF . str_repeat(' ', 28) . $additionalKey;
						$description .= LF;
						foreach ($additionalValue as $ak => $av) {
							$description .= str_repeat(' ', 30) . $ak . ': ' . $av . LF;
						}
					} else {
						$description .= LF . str_repeat(' ', 28) . $additionalKey . ': ' . $additionalValue;
					}
				}
			} else {
				$description = wordwrap($outputValue, self::MAXIMUM_LINE_LENGTH - 28, PHP_EOL . str_repeat(' ', 28), TRUE);
			}
			$this->outputLine('%-2s%-25s %s', array(' ', $outputKey, $description));
		}
	}

	/**
	 * Update list
	 *
	 * Update the list of available extensions in the TER
	 *
	 * @return void
	 * @example ./cli_dispatch.phpsh coreapi extension:updatelist
	 */
	public function extensionUpdatelistCommand() {
		$extensionApiService = $this->getExtensionApiService();
		$extensionApiService->updateMirrors();
		$this->outputLine('Extension list has been updated.');
	}

	/**
	 * List all installed (loaded) extensions
	 *
	 * @param string $type Extension type, can either be L for local, S for system or G for global. Leave it empty for all
	 * @return void
	 * @example ./cli_dispatch.phpsh coreapi extension:listinstalled --type=S
	 */
	public function extensionListinstalledCommand($type = '') {
		$extensionApiService = $this->getExtensionApiService();
		$extensions = $extensionApiService->getInstalledExtensions($type);
		$out = array();

		foreach ($extensions as $key => $details) {
			$title = $key . ' - ' . $details['version'] . '/' . $details['state'];
			$out[$title] = $details['title'];
		}
		$this->outputTable($out);
	}

	/**
	 * Install(activate) an extension
	 *
	 * @param string $key extension key
	 * @return void
	 */
	public function extensionInstallCommand($key) {
		$extensionApiService = $this->getExtensionApiService();
		$data = $extensionApiService->installExtension($key);
		$this->outputLine(sprintf('Extension "%s" is now installed!', $key));
	}

	/**
	 * UnInstall(deactivate) an extension
	 *
	 * @param string $key extension key
	 * @return void
	 */
	public function extensionUninstallCommand($key) {
		$extensionApiService = $this->getExtensionApiService();
		$data = $extensionApiService->uninstallExtension($key);
		$this->outputLine(sprintf('Extension "%s" is now uninstalled!', $key));
	}

	/**
	 * Configure an extension
	 *
	 * This command enables you to configure an extension.
	 *
	 * examples:
	 *
	 * [1] Using a standard formatted ini-file
	 * ./cli_dispatch.phpsh coreapi extension:configure rtehtmlarea --configfile=C:\rteconf.txt
	 *
	 * [2] Adding configuration settings directly on the command line
	 * ./cli_dispatch.phpsh coreapi extension:configure rtehtmlarea --settings="enableImages=1;allowStyleAttribute=0"
	 *
	 * [3] A combination of [1] and [2]
	 * ./cli_dispatch.phpsh extbase extension:configure rtehtmlarea --configfile=C:\rteconf.txt --settings="enableImages=1;allowStyleAttribute=0"
	 *
	 * @param string $key extension key
	 * @param string $configfile path to file containing configuration settings. Must be formatted as a standard ini-file
	 * @param string $settings string containing configuration settings separated on the form "k1=v1;k2=v2;"
	 * @return void
	 * @throws InvalidArgumentException
	 */
	public function extensionConfigureCommand($key, $configfile = '', $settings = '') {
		global $TYPO3_CONF_VARS;
		$extensionApiService = $this->getExtensionApiService();
		$conf = array();

		if (is_file($configfile)) {
			$conf = parse_ini_file($configfile);
		}

		if (strlen($settings)) {
			$arr = explode(';', $settings);
			foreach ($arr as $v) {
				if (strpos($v, '=') === FALSE) {
					throw new InvalidArgumentException(sprintf('Ill-formed setting "%s"!', $v));
				}
				$parts = t3lib_div::trimExplode('=', $v, FALSE, 2);

				if (!empty($parts[0])) {
					$conf[$parts[0]] = $parts[1];
				}
			}
		}

		if (empty($conf)) {
			throw new InvalidArgumentException(sprintf('No configuration settings!', $key));
		}

		$extensionApiService->configureExtension($key, $conf);
		$this->outputLine(sprintf('Extension "%s" has been configured!', $key));

	}


	/**
	 * Fetch an extension from TER
	 *
	 * @param string $key extension key
	 * @param string $version the exact version of the extension, otherwise the latest will be picked
	 * @param string $location where to put the extension. S = typo3/sysext, G = typo3/ext, L = typo3conf/ext
	 * @param string|bool $overwrite overwrite the extension if it already exists
	 * @param string $mirror mirror to fetch the extension from, otherwise a random mirror will be selected
	 * @return void
	 */
	public function extensionFetchCommand($key, $version = '', $location = 'L', $overwrite = FALSE, $mirror = '') {
		$extensionApiService = $this->getExtensionApiService();
		$data = $extensionApiService->fetchExtension($key, $version, $location, $overwrite, $mirror);
		$this->outputLine(sprintf('Extension "%s" version %s has been fetched from repository!', $data['extKey'], $data['version']));
	}


	/**
	 * Import extension from file
	 *
	 * @param string $file path to t3x file
	 * @param string $location where to import the extension. S = typo3/sysext, G = typo3/ext, L = typo3conf/ext
	 * @param boolean $overwrite overwrite the extension if it already exists
	 * @return void
	 */
	public function extensionImportCommand($file, $location = 'L', $overwrite = FALSE) {
		$extensionApiService = $this->getExtensionApiService();
		$data = $extensionApiService->importExtension($file, $location, $overwrite);
		$this->outputLine(sprintf('Extension "%s" has been imported!', $data['extKey']));
	}

	/**
	 * Ensure upload folders of installed extensions exist
	 * @return void
	 */
	public function extensionCreateuploadfoldersCommand() {
		$extensionApiService = $this->getExtensionApiService();
		$messages = $extensionApiService->createUploadFolders();
		if (sizeof($messages)) {
			foreach ($messages as $message) {
				$this->outputLine($message);
			}
		} else {
			$this->outputLine('no uploadFolder created');
		}
	}

	/**
	 * Site info
	 *
	 * Basic information about the system
	 *
	 * @return void
	 * @example ./cli_dispatch.phpsh coreapi site:info
	 */
	public function siteInfoCommand() {
		$siteApiService = $this->getSiteApiService();
		$infos = $siteApiService->getSiteInfo();
		$this->outputTable($infos);
	}

	/**
	 * Create a sys news
	 *
	 * Sys news record is displayed at the login page
	 *
	 * @param string $header Header text
	 * @param string $text Basic text
	 * @return void
	 * @example ./cli_dispatch.phpsh coreapi site:createsysnews "The header" "The news text"
	 */
	public function siteCreatesysnewsCommand($header, $text) {
		$siteApiService = $this->getSiteApiService();
		$siteApiService->createSysNews($header, $text);
	}

	/**
	 * Output a single line
	 *
	 * @param string $text text
	 * @param array $arguments optional arguments
	 * @return void
	 */
	protected function outputLine($text, array $arguments = array()) {
		if ($arguments !== array()) {
			$text = vsprintf($text, $arguments);
		}
		$this->cli_echo($text . PHP_EOL);
	}

	/**
	 * Output a whole table, maximum 2 cols
	 *
	 * @param array $input input table
	 * @return void
	 */
	protected function outputTable(array $input) {
		$this->outputLine(str_repeat('-', self::MAXIMUM_LINE_LENGTH));
		foreach ($input as $key => $value) {
			$line = wordwrap($value, self::MAXIMUM_LINE_LENGTH - 43, PHP_EOL . str_repeat(' ', 43), TRUE);
			$this->outputLine('%-2s%-40s %s', array(' ', $key, $line));
		}
		$this->outputLine(str_repeat('-', self::MAXIMUM_LINE_LENGTH));
	}

	/**
	 * End call
	 *
	 * @param string $message Error message
	 * @return void
	 */
	protected function error($message) {
		$this->cli_echo('ERROR: ' . $message, FALSE);
		die();
	}

	/**
	 * Display help
	 * Overridden from parent
	 *
	 * @return void
	 */
	public function cli_help() {
		if (isset($this->cli_args['_DEFAULT'][1]) &&
				$this->cli_args['_DEFAULT'][1] === 'help' &&
				isset($this->cli_args['_DEFAULT'][2]) &&
				strpos($this->cli_args['_DEFAULT'][2], ':') !== FALSE
		) {
			$this->setHelpFromDocComment($this->cli_args['_DEFAULT'][2]);
		} else {
			$class = new ReflectionClass(get_class($this));
			$commands = array();
			foreach ($class->getMethods(ReflectionMethod::IS_PUBLIC) as $method) {
				if (preg_match($this->commandMethodPattern, $method->getName(), $matches)) {
					$commands[] = strtolower($matches[1]) . ':' . strtolower($matches[2]);
				}
			}
			$this->cli_help['description'] = str_replace('###COMMANDS###', implode(PHP_EOL, $commands), $this->cli_help['description']);
		}
		parent::cli_help();
	}

	/**
	 * Extract help texts from doc comments
	 *
	 * @param $operation
	 */
	protected function setHelpFromDocComment($operation) {
		list($service, $command) = explode(':', $operation, 2);
		$commandMethod = strtolower($service) . ucfirst($command) . 'Command';
		if (method_exists($this, $commandMethod)) {
			$this->cli_help['options'] = '';
			//extract doc comment
			$ref = new ReflectionMethod(get_class($this), $commandMethod);
			$comment = $ref->getDocComment();
			$comment = preg_replace('/\/\*\*\s*(.*)\s*\*\//s', '$1', $comment);
			$lines = explode(PHP_EOL, $comment);

			//get name
			$name = preg_replace('/^\s*\*\s*(.*)\s*$/i', '$1', array_shift($lines));
			$this->cli_help['name'] = $name;

			$description = array();
			$examples = array();
			$params = array();
			foreach ($lines as $n => $l) {
				if (!preg_match('/^\s*\*\s*@/i', $l)) {
					//add to description
					$description[] = preg_replace('/^\s*\*\s*(.*)\s*$/i', '$1', $l);
					continue;
				}

				//params
				if (preg_match('/^\s*\*\s*@param\s*(?P<type>[a-z0-9_]*)\s+\$(?P<name>[a-z0-9_]*)\s+(?P<description>.*)/i', $l, $matches)) {
					$params[$matches['name']] = array(
						'type' => $matches['name'],
						'description' => $matches['description']
					);
					continue;
				}

				// examples
				if (preg_match('/^\s*\*\s*@example\s+(?P<text>.*)/i', $l, $matches)) {
					$examples[] = $matches['text'];
				}
			}

			$this->cli_help['description'] = trim(implode(PHP_EOL, $description));
			if (!empty($examples)) {
				$this->cli_help['examples'] = implode(PHP_EOL, $examples);
			} else {
				unset($this->cli_help['examples']);
			}

			//get params
			$parameters = $ref->getParameters();
			$args = array();
			foreach ($parameters as $param) {
				$name = $param->getName();
				$description = isset($params[$name]) ? $params[$name]['description'] : '';
				if ($param->isOptional()) {
					$this->cli_options[] = array('--' . $name, $description);
				} else {
					$args[strtoupper($name)] = $description;
				}
			}

			//compile arguments section
			if (!empty($args)) {
				$maxLen = 0;
				foreach (array_keys($args) as $argname) {
					if (strlen($argname) > $maxLen) {
						$maxLen = strlen($argname);
					}
				}

				$tmp = array();
				foreach ($args as $argname => $description) {
					$tmp[] = $argname . substr($this->cli_indent(rtrim($description), $maxLen + 4), strlen($argname));
				}

				$offset = array_search('options', array_keys($this->cli_help));
				$this->cli_help = array_slice($this->cli_help, 0, $offset, true) +
						array('arguments' => LF . implode(LF, $tmp)) +
						array_slice($this->cli_help, $offset, NULL, true);
			}

			//set synopsis for this
			$this->cli_help['synopsis'] = './cli_dispatch.phpsh coreapi ' . $operation . ' ###OPTIONS### ' . implode(' ', array_keys($args));

		} else {
			$this->error(sprintf('No help available for "%s"', $operation) . PHP_EOL);
		}
	}

	/**
	 * @return Tx_Coreapi_Service_CacheApiService
	 */
	protected function getCacheApiService() {
		$cacheApiService = t3lib_div::makeInstance('Tx_Coreapi_Service_CacheApiService');
		$cacheApiService->initializeObject();
		return $cacheApiService;
	}

	/**
	 * @return Tx_Coreapi_Service_DatabaseApiService
	 */
	protected function getDatabaseApiService() {
		$databaseApiService = t3lib_div::makeInstance('Tx_Coreapi_Service_DatabaseApiService');
		return $databaseApiService;
	}

	/**
	 * @return Tx_Coreapi_Service_ExtensionApiService
	 */
	protected function getExtensionApiService() {
		$extensionApiService = t3lib_div::makeInstance('Tx_Coreapi_Service_ExtensionApiService');
		return $extensionApiService;
	}

	/**
	 * @return Tx_Coreapi_Service_SiteApiService
	 */
	protected function getSiteApiService() {
		$siteApiService = t3lib_div::makeInstance('Tx_Coreapi_Service_SiteApiService');
		return $siteApiService;
	}
}

?>